dev/com.ibm.ws.threading/src/com/ibm/ws/threading/internal/ThreadPoolController.java

```diff
@@ -1504,16 +1512,23 @@ private boolean resolveHang(long tasksCompleted, boolean queueEmpty, int poolSiz
                 Tr.event(tc, "Executor hang continued at poolSize=" + poolSize, threadPool);
             }
 
-            setPoolIncrementDecrement(poolSize);
-            if (poolSize + poolIncrement <= maxThreads && poolSize < MAX_THREADS_TO_BREAK_HANG) {
-                targetPoolSize = adjustPoolSize(poolSize, poolIncrement);
+            /**
+             * We might get here with a prior pool size change incomplete ... if so we
+             * should adjust for that
+             */
+            int hangSizeAdjustment = 0;
+            if (poolSize > targetPoolSize) {
+                setPoolIncrementDecrement(targetPoolSize);
+                hangSizeAdjustment = poolIncrement - (poolSize - targetPoolSize);
+            } else if (poolSize < targetPoolSize) {
+                hangSizeAdjustment = targetPoolSize - poolSize;
+            }
+            setPoolIncrementDecrement(poolSize + hangSizeAdjustment);
+            int adjustedPoolIncrement = hangSizeAdjustment == 0 ? poolIncrement : hangSizeAdjustment;
+            if (poolSize + adjustedPoolIncrement <= Math.min(maxThreads, MAX_THREADS_TO_BREAK_HANG)) {
```

Comment:
This really changes the meaning.  Before we were able to adjust the thread pool to a higher size that was over max threads to break hang and then the next time it would not allow something higher.  Now if we determine to do something higher that is over the max threads to break hang, but the previous value is below it, we will not allow to go higher to break the hang.  As such max threads to break the hang is not the max in reality because it may be less than that that we start rejecting.